cpp/cmake_modules/BuildUtils.cmake

```diff
@@ -154,22 +161,28 @@ function(ADD_ARROW_LIB LIB_NAME)
   endif()
 
   if (ARROW_BUILD_STATIC)
-      if (MSVC)
-        set(LIB_NAME_STATIC ${LIB_NAME}_static)
-      else()
-        set(LIB_NAME_STATIC ${LIB_NAME})
-      endif()
-      add_library(${LIB_NAME}_static STATIC $<TARGET_OBJECTS:${LIB_NAME}_objlib>)
+    if (MSVC)
+      set(LIB_NAME_STATIC ${LIB_NAME}_static)
+    else()
+      set(LIB_NAME_STATIC ${LIB_NAME})
+    endif()
+    add_library(${LIB_NAME}_static STATIC ${LIB_DEPS})
+    if(EXTRA_DEPS)
+      add_dependencies(${LIB_NAME}_static ${EXTRA_DEPS})
+    endif()
+
     set_target_properties(${LIB_NAME}_static
       PROPERTIES
       LIBRARY_OUTPUT_DIRECTORY "${BUILD_OUTPUT_ROOT_DIRECTORY}"
       OUTPUT_NAME ${LIB_NAME_STATIC})
 
-  target_link_libraries(${LIB_NAME}_static
+    target_compile_definitions(${LIB_NAME}_static PUBLIC ARROW_STATIC)
```

Comment:
@JohnPJenkins  To avoid confusion, it might make sense to define this only for MSVC